scopes/preview/preview/pre-bundle-utils.ts

```diff
@@ -0,0 +1,75 @@
+import { join } from 'path';
+import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs-extra';
+import { UIRoot } from '@teambit/ui';
+import { getAspectDirFromBvm } from '@teambit/aspect-loader';
+import { SlotRegistry } from '@teambit/harmony';
+import { ArtifactDefinition } from '@teambit/builder';
+import { sha1 } from '@teambit/legacy/dist/utils';
+
+/// utils
+
+export type UIRootRegistry = SlotRegistry<UIRoot>;
+
+// bundle hash
+
+export const BUNDLE_HASH_FILENAME = '.hash';
+
+export function readBundleHash(bundleId: string, bundleDir: string, aspectDir: string): string {
+  const bundleUiPathFromBvm = getBundlePath(bundleId, bundleDir, aspectDir);
+  if (!bundleUiPathFromBvm) {
+    return '';
+  }
+  const hashFilePath = join(bundleUiPathFromBvm, BUNDLE_HASH_FILENAME);
+  if (existsSync(hashFilePath)) {
+    return readFileSync(hashFilePath).toString();
+  }
+  return '';
+}
+
+export async function createBundleHash(uiRoot: UIRoot, runtime: string): Promise<string> {
+  const aspects = await uiRoot.resolveAspects(runtime);
+  aspects.sort((a, b) => ((a.getId || a.aspectPath) > (b.getId || b.aspectPath) ? 1 : -1));
+  const aspectPathStrings = aspects.map((aspect) => {
+    return [aspect.aspectPath, aspect.runtimePath].join('');
+  });
+  return sha1(aspectPathStrings.join(''));
+}
+
+export async function generateBundleHash(uiRoot: UIRoot, runtime: string, outputPath: string): Promise<void> {
+  const hash = await createBundleHash(uiRoot, runtime);
+  if (!existsSync(outputPath)) mkdirSync(outputPath);
+  writeFileSync(join(outputPath, BUNDLE_HASH_FILENAME), hash);
```

Comment:
Yeah, I would prefer to refactor ui.main.runtime as a separated PR if you don't mind. And the refactoring is not only about these utils. Also because we met a lot of webpack compilation errors when last time when I coupled those 2 packages too much. What do you think?